_documentation/Recipes/InlineIPsec.md

```diff
@@ -0,0 +1,63 @@
+
+![Top Level Components](../img/InlineIPsec_Recipe.svg "Inline IPsec Top Level"){:width="60%"}
+
+# Infrastructure Application Interface #
+
+ - [strongSwan](https://www.strongswan.org/) is used as the IKE control plane.  No code changes are needed in strongSwan.
+ - A strongSwan plug-in is used to convert security association (SA) information into OpenConfig RPC messages to the underlying infrastructure.  The P4 program defines which modes are required in the pipeline, as well as where the IPsec is inserted into the broader virtual networking pipeline:
+   - **Tunnel vs. Transport**: Determines how the original packet is transformed during encryption/decryption as below: 
+![IPsec Packet Transforms](../img/IPsec_Packet_Transforms.svg "IPsec Packet Transforms"){:width="60%"}
+   - **Encrypted Virtual Ports**: Encrypted tunnels between devices over physical ports.  On the wire this wraps additional headers on the outside of the packet in the same way that network virtualization headers such as VXLAN are added/removed.
+   - **Encrypted Physical Ports**: Encrypted virtual ports plugged into compute instances (all traffic going in/out of a given virtual port will be encrypted/decrypted).  On the wire this puts the IPsec headers onto the innermost headers belonging to the end user of the compute instance.
+
+
+The use of cryptography to secure data plane traffic has become common practice. Encrypting and decrypting network traffic using host resources requires that the host must also perform any packet processing of post-decryption receive packets and pre-encryption transmit packets. This can consume significant host compute resources. Offloading packet crypto processing to the data plane removes the burden for cryptographic processing from the host and also allows the infrastructure to be used for pre-encryption and post-decryption packet processing functions.
+
+This recipe looks at inline IPsec processing where the crypto is terminated in the infrastructure, which avoids having specialized drivers or control planes on the compute instance(s). The IPsec control plane (IKE protocol) is offloaded to the infrastructure. The IKE protocol processing runs in software on the infrastructure cores and programs the data plane with offloaded IPsec flows. The data plane programming is separated into two parts, P4Runtime for programming the P4 pipeline and OpenConfig for programming the IPsec crypto configuration. The IPsec crypto configuration includes the IPsec Security Association (SA) table entries (i.e. crypto keys and algorithm selection) and is used for reporting events related to lifetime expiration of SA entries.
```

Comment:
I assume the IPsec crypto configuration also does rekeying and re-auth. Is that also handled over OpenConfig somehow?

Reply:
yes I added rekey and re-auth in both the InfraApp and TAI interfaces